paimon-core/src/main/java/org/apache/paimon/mergetree/compact/LoserTree.java

```diff
@@ -0,0 +1,347 @@
+                T childKey = winnerNode.peek();
+                int secondResult = secondComparator.compare(parentKey, childKey);
+                if (secondResult > 0) {
+                    parentNode.state = State.WINNER_WITH_SAME_KEY;
+                    winnerNode.state = State.LOSER_WITH_SAME_KEY;
+                    parentNode.setFirstSameKeyIndex(index);
+                } else {
+                    winnerNode.setFirstSameKeyIndex(index);
+                }
+                return;
+            case LOSER_WITH_NEW_KEY:
+            case LOSER_POPPED:
+                return;
+            default:
+                throw new UnsupportedOperationException(
+                        "unknown state for " + parentNode.state.name());
+        }
+    }
+
+    /**
+     * The userKey of the new local winner node is different from that of the previous global
+     * winner.
+     */
+    private void adjustWithNewWinnerKey(
+            int index, LeafIterator<T> parentNode, LeafIterator<T> winnerNode) {
+        switch (parentNode.state) {
+            case LOSER_WITH_NEW_KEY:
+                // when the new winner is also a new key, it needs to be compared.
+                T parentKey = parentNode.peek();
+                T childKey = winnerNode.peek();
+                int firstResult = firstComparator.compare(parentKey, childKey);
+                if (firstResult == 0) {
+                    // if the compared keys are the same, we need to update the state of the node
+                    // and record the index of the same key for the winner.
+                    int secondResult = secondComparator.compare(parentKey, childKey);
+                    if (secondResult < 0) {
+                        parentNode.state = State.LOSER_WITH_SAME_KEY;
+                        winnerNode.setFirstSameKeyIndex(index);
+                    } else {
+                        winnerNode.state = State.LOSER_WITH_SAME_KEY;
+                        parentNode.state = State.WINNER_WITH_NEW_KEY;
+                        parentNode.setFirstSameKeyIndex(index);
+                    }
+                } else if (firstResult > 0) {
+                    // the two keys are completely different and just need to update the state.
+                    parentNode.state = State.WINNER_WITH_NEW_KEY;
+                    winnerNode.state = State.LOSER_WITH_NEW_KEY;
+                }
+                return;
+            case LOSER_WITH_SAME_KEY:
+                // the previous loser has the same key as the winner of the previous time,
+                // so the comparison is skipped directly.
+                parentNode.state = State.WINNER_WITH_SAME_KEY;
+                winnerNode.state = State.LOSER_WITH_NEW_KEY;
+                return;
+            case LOSER_POPPED:
+                parentNode.state = State.WINNER_POPPED;
+                parentNode.firstSameKeyIndex = -1;
+                winnerNode.state = State.LOSER_WITH_NEW_KEY;
+                return;
```

Comment:
This will only happen during `adjustForNextLoop`, not during `popWinner`?

Reply:
Yes, because the local winner node is in `WINNER_WITH_NEW_KEY` state, this will only happen in `initialize` and `adjustForNextLoop`.

Reply:
Add comments in code stating that this will only happen during `adjustForNextLoop`.